clippy_lints/src/unnecessary_wraps.rs

```diff
@@ -110,39 +133,42 @@ impl<'tcx> LateLintPass<'tcx> for UnnecessaryWraps {
         });
 
         if can_sugg && !suggs.is_empty() {
-            span_lint_and_then(
-                cx,
-                UNNECESSARY_WRAPS,
-                span,
-                format!(
-                    "this function's return value is unnecessarily wrapped by `{}`",
-                    return_type
-                )
-                .as_str(),
-                |diag| {
-                    let inner_ty = return_ty(cx, hir_id)
-                        .walk()
-                        .skip(1) // skip `std::option::Option` or `std::result::Result`
-                        .take(1) // take the first outermost inner type
-                        .filter_map(|inner| match inner.unpack() {
-                            GenericArgKind::Type(inner_ty) => Some(inner_ty.to_string()),
-                            _ => None,
-                        });
-                    inner_ty.for_each(|inner_ty| {
+            // Issue 6640: If the inner type is Unit, emit lint similar to clippy::unused_unit.
+            if inner_ty.is_unit() {
+                span_lint_and_sugg(
```

Comment:
This is basically a copy from `unused_unit`. Should I change it?

Reply:
You should probably use `span_lint_and_then` here as well because this lint suggests changes at several places:
1. Changing the return type
2. Changing the actual return expression

`span_lint_and_sugg` only works for one suggestion this would leave the return statements unchanged which would lead to syntax errors until the user updates them. `span_lint_and_then` allows a suggesting for the return type and the for the expressions using `multipart_suggestion`.

--- 

I personally would try to reuse the `span_lint_and_then` and just adjust the suggestions and message string according to the case.
1. We change or remove the return type (`type_snippet`)
2. We have a message which is changed according to the suggestion (`type_change`)
3. The generated suggestions are also adapted based on the change so either 
    * `return Some(XYZ);` -> `return XYZ;` (`Some(XYZ)` -> `XYZ`)
    * `return Some(());` -> `return;` (`Some(())` -> ``)

This would allow the `span_lint_and_then` reuse like this:

```rust
let (type_change, type_snippet) = if inner_ty.is_unit() {
    (
        format!("Remove the return declaration like this:"),
        String::new(),
    )
} else {
    (
        format!("remove `{}` from the return type like this:", return_type_label),
        return_type_label.to_string(),
    )
};

let msg = format!("this function's return value is unnecessarily wrapped by `{}`", return_type_label);

span_lint_and_then(
    cx,
    UNNECESSARY_WRAPS,
    span,
    msg.as_str(),
    |diag| {
        diag.span_suggestion(
            fn_decl.output.span(),
            type_change.as_str(),
            type_snippet,
            Applicability::MaybeIncorrect,
        );
        diag.multipart_suggestion(
            "And then change the returning expressions like this:",
            suggs,
            Applicability::MaybeIncorrect,
        );
    },
);
```

That's just my own style, so there might be better ways or some other way which seems cleaner to you. Just use the one you are most comfortable with :upside_down_face:

Reply:
Thank you for the great suggestion. However, there is one thing I am not certain with. One issue I see with using `span_lint_and_then()` for cases like `-> Option<()>`  is that depending on the function, we would have two different kinds of suggestions: 1) `return ();` -> `return;` and 2) `()` -> remove the line. I wonder if we can always propose good suggestions in both cases, especially for 2).

Reply:
We should still notify the user even if they should just remove a line. `multipart_suggestion` can also be used to remove code by just suggesting an empty string. You already use this in the replacement of the `-> Option<()>`. The only thing you would need to adapt is the suggestion generation (Here is the change in pseudo rust):

```rust
// line 110

// Check if all return expression respect the following condition and collect them.
let is_inner_union = inner_ty.is_unit();
let mut suggs = Vec::new();
let can_sugg = find_all_ret_expressions(cx, &body.value, |ret_expr| {
    if_chain! {
        [...]
        then {
            if is_inner_union {
                suggs.push((ret_expr.span, "".to_string()));
            } else {
                suggs.push((ret_expr.span, snippet(cx, args[0].span.source_callsite(), "..").to_string()));
            }
            true
        } else {
            false
        }
    }
});
```

The span is just set to `ret_expr.span`. I **believe** that this would only be the `Some(())` part from `return Some(())`. The suggestion would therefore leave the simple return, and we don't need to handle any special cases. Could you maybe try this? :upside_down_face:

Reply:
Thanks again for the great suggestion. I had already started to try that, but I had given up for the above reason.
So it works, but I see two potential issues:
1. For the case with a `return`, it works, but since the white space between it and the return expression is not included in the latter span, I had to hack a function to included it. I works for my tests, but I don't know if it's robust enough:
```rust
// Given an expression, if it parents expression is a return expression, return its span with the
// whitespaces after `return` keyword. For example, if we have `return 1;`, by passing the
// expression `1`, this function would return the span of ` 1`.
pub fn get_ret_expr_span_with_prefix_space(cx: &LateContext<'_>, ret_expr: &Expr<'_>, span: Span) -> Option<Span> {
	if_chain! {
		// Try to get parent expression which would include `return`.
		if let Some(parent_expr) = get_parent_expr(cx, ret_expr);
		// Verify if that expression is ExprKind::Ret.
		if let ExprKind::Ret(_) = parent_expr.kind;
		// Get the parent expression source.
		if let Ok(parent_expr_source) = cx
			.sess()
			.source_map()
			.span_to_snippet(span.with_lo(parent_expr.span.lo()).with_hi(parent_expr.span.hi()));
		// Find "return"
		if let Some(pos) = parent_expr_source.find("return");
		then {
			#[allow(clippy::cast_possible_truncation)]
			Some(parent_expr.span.with_lo(BytePos(parent_expr.span.lo().0 + pos as u32 + 6)))
		} else {
			None
		}
	}
}
```
2. As you said, the empty string does remove the `Some(())`, but it leaves an empty line (see between `Some(());` and `} else {:
```
LL |         return;
LL |     }
LL |     if a {
LL |         Some(());
LL |
LL |     } else {
``` 
Is there a way to remove this? Or is it fine like this?

Reply:
Creating such a function is a valid option and the more I think about it probably also the cleanest option. :+1: 

I would suggest adding it as an util function similar to [`utils::position_before_rarrow`](https://github.com/rust-lang/rust-clippy/blob/2f19f5f0d71833d450f8290bb154c3572ac58d9d/clippy_lints/src/utils/mod.rs#L761-L787). This is probably something that will be reused in the future. You can then use this by 
1. Calling [`utils::line_span`](https://github.com/rust-lang/rust-clippy/blob/2f19f5f0d71833d450f8290bb154c3572ac58d9d/clippy_lints/src/utils/mod.rs#L789-L804) with the expression span.
2. Snipping the returned span. This should then be the entire line as a string.
3. Get the offset with the new util function.
4. Add the offset to the line span.

All of these steps can probably also be done by the util function itself. 

If you want to remove the entire line like you wanted in your example output you can simply use  [`utils::line_span`](https://github.com/rust-lang/rust-clippy/blob/2f19f5f0d71833d450f8290bb154c3572ac58d9d/clippy_lints/src/utils/mod.rs#L789-L804). The visualization will most likely still show the empty line though. The lint has multiple spans with suggestions and is therefor not automatically applicable anyways. The suggestion is therefor only for the user to see which changes should be done :).

Another way would be to suggest a change that includes the second to last statement like this:

```rust
    if {
        println!("Hello world!");
// From here    v
        Some(());
        return Option::Some(());
    }
//  ^ To here
```

The suggestion would then show the suggestion like:
```rust
LL |        Some(());
LL |    }
```

The suggestion is a bit cleaner but the code to get this suggestion a more complicated with more edge cases. These are the options I can think of. It might be worth to ask over on our [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy) channel if you want more suggestions :).

Reply:
Thank you again for the long reply.
To be honest, I think I prefer leaving an empty line. I am also worried that by trying to remove line, we might hit some weird edge cases. Just removing the the span of the return expression seems much safer even if the suggestion will not be as clean. So if you approve, I'll go this way.
About the utility function, a function `position_after_return()` would just be a call to `find()`, so I'm not sure it would be worth it. I could add this whole function, but I feel it is a bit too sketchy to go in utisl. :-P But you tell me. :-)

Reply:
Generally using `ByteOffset`s in suggestions is almost always a bad idea. That is because it almost always assumes a specific formatting of the code. Also it may remove comments. 

In this case I would just suggest to remove the `Some(())` and just keep the empty line. If the user cares about formatting, `rustfmt` will deal with that anyway. If not ¯\\\_(ツ)\_/¯. Formatting shouldn't be a concern for Clippy.

---

I also thought about adding a utils function, that produces a suggestion to remove an entire line before. But then I found this case:

```rust
if xyz { Some(()) } else { Err(()) }
```

which is valid formatting according to `rustfmt` with some configurations. Just removing the entire line or everything `rfind("\n")` would break the code.

Reply:
Thank you for the comment.
Yes, I was a bit uncomfortable with using `ByteOffset`. However, without it, I don't know if there is a way to avoid suggesting replacing `return Some(());` by `return ;`. That white space it annoying. :-P
Any alternative?

Reply:
Nope, just ignore that whitespace. Formatting shouldn't be a concern for Clippy.